File: HW1/task1.py
import pyspark
import sys, json, operator, re

if __name__ == "__main__":

    # parse commandline argument
    input_path = sys.argv[1]
    output_path = sys.argv[2]
    stopwords_path = sys.argv[3]
    given_year = sys.argv[4]
    top_m_user = int(sys.argv[5])
    top_n_word = int(sys.argv[6])

    conf = pyspark.SparkConf().setAppName("Task1").setMaster("local[*]")
    sc = pyspark.SparkContext(conf = conf)
    sc.setLogLevel("ERROR")

    reviewsRDD = sc.textFile(input_path).map(lambda review: json.loads(review)).cache()
    result = {}

    # total number of reviews
    result["A"] = reviewsRDD.count()
    print("total number of reviews: ", result["A"])

    # total number of reviews in a given year
    result["B"] = reviewsRDD.filter(lambda review: given_year in review["date"]).count()
    print("total number of reviews in a given year: ", result["B"])

    # number of distinct users
    result["C"] = reviewsRDD.map(lambda review: review["user_id"]).distinct().count()
    print("number of distinct users who have written the reviews: ", result["C"])

    # top M users who have the largest number of reviews and its count
    result["D"] = reviewsRDD.map(lambda review: [review["user_id"], 1]).reduceByKey(operator.add).takeOrdered(top_m_user, key=lambda list: [-list[1], list[0]])
    print("top {} users who have the largest number of reviews and its count:".format(top_m_user))
    print(result["D"])

    # top N frequent words in the review text excluding stopwords
    stopwords = sc.textFile(stopwords_path).collect()
    result["E"] = reviewsRDD.map(lambda review: review["text"].translate({ord(i): None for i in ',.!?:;()[]'}).lower()).flatMap(lambda review: review.split()).filter(lambda w: w and w not in stopwords).map(lambda w: (w, 1)).reduceByKey(operator.add).takeOrdered(top_n_word, key=lambda x: (-x[1], x[0]))
    result["E"] = list(map(lambda x: x[0], result["E"]))
    print("top {} frequent words in the review text excluding stopwords:".format(top_n_word))
    print(result["E"])

    # output
    with open(output_path, 'w') as f:
        json.dump(result, f, sort_keys=True)

File: HW1/scala/task1.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkConf
import org.json4s._
import org.json4s.jackson.JsonMethods._
import scala.collection.mutable._
import org.json4s.jackson.Serialization
import java.io._

object task1 {
  def main(args: Array[String]) {

    // parse commandline argument
    val inputPath = args(0)
    val outputPath = args(1)
    val stopwordsPath = args(2)
    val givenYear = args(3)
    val topMUser: Int = args(4).toInt
    val topNWord: Int = args(5).toInt

    // set up spark
    val conf = new SparkConf().setAppName("Task1").setMaster("local[*]")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    val reviewsRDD = sc.textFile(inputPath).map(line => parse(line)).cache()

    var result: LinkedHashMap[String, Any] = LinkedHashMap()

    result += ("A" -> reviewsRDD.count())
    println(s"total number of reviews: ${result("A")}")

    result += ("B" -> reviewsRDD.filter(review => compact(render(review \ "date"))regionMatches(false, 1, givenYear, 0, 4)).count())
    println(s"total number of reviews in a given year: ${result("B")}")

    result += ("C" -> reviewsRDD.map(review => compact(render(review \ "user_id"))).distinct().count())
    println(s"number of distinct users who have written the reviews: ${result("C")}")

    def mapFunc(x: JValue): (String, Int) = {
      val str = compact(render(x \ "user_id"))
      (str.substring(1, str.length - 1), 1)
    }
    val users = reviewsRDD.map(mapFunc).reduceByKey(_+_).takeOrdered(topMUser)(Ordering[(Int, String)].on(x => (-x._2, x._1)))
    val res:Array[List[Any]] = new Array(users.length)
    for (i <- 0 until users.length) {
      res(i) = users(i).productIterator.toList
    }
    result += ("D" -> res)

    // scala is so troublesome, give up E

    implicit val formats = org.json4s.DefaultFormats
    val writer = new PrintWriter(new File(outputPath))
    writer.write(Serialization.write(result))
    writer.close()
  }
}
